### PR TITLE
feat(dataset): dataset create and add using new metadata

### LINKED ISSUE #1735: Add dataset commands for new graph metadata structure
Add all dataset commands under `renku graph dataset` (e.g. `renku graph dataset add`), but using the new metadata structure, keeping the old behavior.

### PR DESCRIPTION
**Pull request by mohammad-sdsc**
# Description

Use new metadata in `graph dataset create` and `graph dataset add` commands.

Fixes #1735

### CODE REVIEW

**Comment by mohammad-sdsc** [comment]
Loading all dataset provenance is not needed frequently when working with datasets. In common case, we only need to load the latest version of metadata for a dataset and make some modifications in it. Therefore, loading the whole history is not needed and only adds to data loading time.

File: renku/core/models/provenance/datasets.py [hunk 1]
```diff
@@ -226,6 +269,20 @@ def from_dataset(cls, dataset, client, revision):
...
         )
 
+    @staticmethod
+    def _convert_from_dataset_files(files, client, revision):
+        dataset_files = []
+        files = {f.path: f for f in files}  # NOTE: To make sure there are no duplicate paths
+        for path in files:
+            file = files[path]
+            dataset_file = DatasetFile.from_dataset_file(file, client=client, revision=revision)
+            if not revision:
```
Comment (Panaetius) [inline renku/core/models/provenance/datasets.py:279]:
```suggestion
            if not dataset_file:
```

File: renku/core/models/provenance/datasets.py [hunk 2]
```diff
@@ -377,13 +443,30 @@ def add(self, dataset):
...
+            if dataset.derived_from:
+                datasets_ids.remove(dataset.derived_from.url_id)
+
+        assert len(datasets_ids) <= 1, f"There are more than one latest versions with name `{name}`"
+
+        if not datasets_ids:
+            return None
+
+        url = datasets_ids.pop()
+        return next(d for d in self.get_by_name(name) if d.id == url)
```
Comment (Panaetius) [inline renku/core/models/provenance/datasets.py:469]:
```suggestion
        datasets = {d.id: d for d in self.get_by_name(name)}

        for dataset in datasets.values():
            if dataset.derived_from:
                datasets.pop(dataset.derived_from.url_id, None)

        assert len(datasets) <= 1, f"There are more than one latest versions with name `{name}`"

        if not datasets:
            return None

        _, dataset = datasets.popitem()
        return dataset
```

would reduce the number of times we need to iterate over this.

File: tests/cli/test_datasets.py [hunk 3]
```diff
@@ -521,7 +521,8 @@ def test_repository_file_to_dataset(runner, project, client, subdirectory):
     client.repo.git.commit(message="Added file a", no_verify=True)
 
     # add data
-    result = runner.invoke(cli, ["dataset", "add", "dataset", str(a_path)], catch_exceptions=False,)
+    command = ["graph"] if use_graph else []
```
Comment (Panaetius) [inline tests/cli/test_datasets.py:524]: Cool idea!

**Review by Panaetius — CHANGES_REQUESTED** [review]

File: renku/core/models/provenance/datasets.py (hunk 2 above)
Comment (mohammad-sdsc) [reply renku/core/models/provenance/datasets.py]: We cannot iterate over a dict and remove items from it; I've just changed `for dataset in datasets.values()` to `for dataset in self.get_by_name(name)`

Comment (Panaetius) [reply renku/core/models/provenance/datasets.py]: right, `for dataset in list(datasets.values())` would work though. I'm just thinking in a project with a large provenance, we want to loop over the provenance as little as possible.

File: renku/cli/exception_handler.py [hunk 4]
```diff
@@ -86,7 +86,7 @@ def main(self, *args, **kwargs):
         try:
             return super().main(*args, **kwargs)
         except RenkuException as e:
-            click.echo("Error: {}".format(e))
+            click.echo("Error: {}".format(e), err=True)
```
Comment (jsam) [inline renku/cli/exception_handler.py:89]: Should we switch this to f-string?

File: renku/core/management/datasets.py [hunk 5]
```diff
@@ -247,6 +250,51 @@ def with_dataset(self, name=None, create=False):
...
+            raise errors.DatasetExistsError('Dataset exists: "{}".'.format(name))
+        else:
+            dataset = dataset.to_dataset(self)
+
+        dataset_path = self.path / self.data_dir / dataset.name
+        dataset_path.mkdir(parents=True, exist_ok=True)
+
+        try:
+            yield dataset
+        except Exception:
```
Comment (jsam) [inline renku/core/management/datasets.py:277]: Should we use `RenkuException` here?

File: renku/core/models/provenance/datasets.py [hunk 6]
```diff
@@ -128,6 +152,25 @@ def _extract_identifier(self):
...
+                revision=revision,
+                added=self.date_added,
+                based_on=self.based_on,
+                external=self.is_external,
+                id=None,
+                path=self.entity.path,
+                source=self.source,
+                url=None,
+            )
+        except KeyError:  # path does not exists at revision
```
Comment (jsam) [inline renku/core/models/provenance/datasets.py:171]: Could we add a tag `NOTE:` and also which line from `.from_revision` is expected to raise this?

File: renku/core/models/provenance/datasets.py [hunk 7]
```diff
@@ -226,6 +269,20 @@ def from_dataset(cls, dataset, client, revision):
             version=dataset.version,
         )
 
+    @staticmethod
+    def _convert_from_dataset_files(files, client, revision):
```
Comment (jsam) [inline renku/core/models/provenance/datasets.py:273]: Should we add a doc string here?

File: renku/core/models/provenance/datasets.py [hunk 8]
```diff
@@ -316,37 +373,46 @@ def _update_files(self, dataset, client, revision, date):
...
+            license=self.license,
+            same_as=self.same_as,
+            tags=self.tags,
+            title=self.title,
+            url=self.url,
+            version=self.version,
         )
 
-        dataset_files.append(dataset_file)
+    def _convert_to_dataset_files(self, client):
```
Comment (jsam) [inline renku/core/models/provenance/datasets.py:406]: Should we add a doc string here?

File: renku/core/models/provenance/datasets.py [hunk 9]
```diff
@@ -377,13 +443,30 @@ def add(self, dataset):
 
     def get(self, identifier):
         """Return a dataset by its original identifier."""
-        datasets = [d for d in self._datasets if d.identifier == identifier]
-        assert len(datasets) <= 1, f"Found more than one with identifier `{identifier}`."
-        return datasets[0] if datasets else None
+        datasets = (d for d in self._datasets if d.identifier == identifier)
+        dataset = next(datasets, None)
+        assert next(datasets, None) is None, f"Found more than one dataset with identifier `{identifier}`."
```
Comment (jsam) [inline renku/core/models/provenance/datasets.py:448]: Should replace this assert with explicit check and raise?

**Review by jsam — COMMENTED** [review]
Thank you, it look's great! 

I just added a few comments/questions, otherwise we can :rocket:

File: renku/core/management/datasets.py (hunk 5 above)
Comment (mohammad-sdsc) [reply renku/core/management/datasets.py:277]: No, this should catch non-Renku exceptions to do a cleanup.

File: renku/core/models/provenance/datasets.py (hunk 9 above)
Comment (mohammad-sdsc) [reply renku/core/models/provenance/datasets.py:448]: We've added these checks as safety-nets here and in other places to make sure that they never happen. I'll keep the format as it is to be in par with other places. End users should not see this (or the repo is really broken).

File: renku/core/models/provenance/datasets.py (hunk 6 above)
Comment (mohammad-sdsc) [reply renku/core/models/provenance/datasets.py]: I've changed the wording to make it clear which part of that function raises this error.

**Review by Panaetius — APPROVED** [review]

**Pull request merged**